{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9afdd150",
   "metadata": {},
   "source": [
    "### Normalization and highly variable genes (HVG)\n",
    "\n",
    "**Objective**\n",
    "Normalize gene expression values and identify highly variable genes for downstream analysis.\n",
    "\n",
    "*How can technical variability be minimized to allow meaningful comparison between cells?* \n",
    "\n",
    "**Input**\n",
    "- Cleaned object from cell quality filtering (Notebook 02)\n",
    "\n",
    "**Methods**\n",
    "- Library-size normalization\n",
    "- Log-transformation\n",
    "- Identification of highly variable genes (HVG)\n",
    "\n",
    "**Output**\n",
    "- Normalized object\n",
    "- List of highly variable genes\n",
    "\n",
    "Normalization and log-transformation were applied to correct for differences in sequencing depth across cells.\n",
    "Highly variable genes were then selected to capture the main sources of biological variability, ensuring that downstream analyses focus on informative signals rather than technical noise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c229166c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory: /mnt/c/Users/yasmi/OneDrive/Desktop/Mini-Projets/scRNA_Influenza_Patients\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "PROJECT_ROOT = \"/mnt/c/Users/yasmi/OneDrive/Desktop/Mini-Projets/scRNA_Influenza_Patients\"\n",
    "os.chdir(PROJECT_ROOT)\n",
    "\n",
    "print(\"Working directory:\", os.getcwd())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7cedce9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yasmina/miniconda3/lib/python3.13/site-packages/anndata/_core/anndata.py:1806: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HVG: 2000\n",
      "AnnData object with n_obs × n_vars = 103202 × 37493\n",
      "    obs: 'gsm', 'sample_id', 'condition', 'replicate', 'batch'\n",
      "    var: 'gene_ids', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'log1p', 'hvg'\n",
      "After HVG subsetting: AnnData object with n_obs × n_vars = 103202 × 2000\n",
      "    obs: 'gsm', 'sample_id', 'condition', 'replicate', 'batch'\n",
      "    var: 'gene_ids', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
      "    uns: 'log1p', 'hvg'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yasmina/miniconda3/lib/python3.13/site-packages/anndata/_core/anndata.py:1806: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Fix project root (comme dans les autres notebooks)\n",
    "os.chdir(\"/mnt/c/Users/yasmi/OneDrive/Desktop/Mini-Projets/scRNA_Influenza_Patients\")\n",
    "\n",
    "# Load QC-filtered data\n",
    "adata = sc.read_h5ad(\"results/adata_with_condition_raw.h5ad\")\n",
    "\n",
    "# 1) Normalisation (rendre les cellules comparables)\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "\n",
    "# 2) Log-transformation\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# 3) Identification des gènes hautement variables (HVG)\n",
    "sc.pp.highly_variable_genes(adata, flavor=\"seurat\", n_top_genes=2000)\n",
    "\n",
    "# Vérifications rapides\n",
    "print(\"HVG:\", adata.var[\"highly_variable\"].sum())\n",
    "print(adata)\n",
    "\n",
    "# Keep only HVGs for downstream analysis\n",
    "adata = adata[:, adata.var[\"highly_variable\"]].copy()\n",
    "print(\"After HVG subsetting:\", adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "acd1a1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"results/adata_norm_hvg.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
